### PR TITLE
fix(cayenne): report the auto-tuned config once per resolution (fixes #12330)

### LINKED ISSUE #12330: Cayenne auto-tuned config line is logged once per provider construction, not once per unique config
### Summary

The `Cayenne auto-tuned config: …` INFO line is emitted once per table-provider construction rather than once per resolved config. Because dataset initialization retries with unbounded backoff, a dataset that never loads re-emits its identical ~1.1 KB config line forever.

Each unique config should be logged once.

### Where

`crates/runtime/src/dataaccelerator/cayenne/mod.rs:1933` — emitted from `get_vortex_config_with_footer_cache`, which is reached on every `create_external_table` / `create_cayenne_table_provider` call. Nothing dedupes on the resolved config, so the line is a function of *how many times a provider was built*, not of *how many distinct configs exist*.

The amplifier is `crates/runtime/src/init/dataset.rs:569`:

```rust
let retry_strategy = FibonacciBackoffBuilder::new().max_retries(None).build();
```

Only `Error::PermanentDatasetFailure` short-circuits; everything else is classified `RetryError::transient`. A deterministic configuration error therefore retries indefinitely, and each attempt reconstructs the provider and re-logs the config.

### Observed

Three Cayenne-accelerated datasets, one of which fails a schema-validation check at load. In the first ~4.5 minutes of runtime uptime the line was emitted **33 times for 3 tables — 3 unique configs**, at fibonacci-spaced intervals (1s, 2s, 3s, 5s, 8s, 13s, 21s, 34s, 55s…), and continues for as long as the process runs.

Two of the three configs are byte-identical to each other; they differ only in `sort_columns`.

```
00:23:05  Cayenne auto-tuned config: segment_cache=256MB, … table=table_a …
00:23:05  Cayenne auto-tuned config: segment_cache=256MB, … table=table_b …
00:23:05  Cayenne auto-tuned config: segment_cache=256MB, … table=table_c …
00:23:06  Cayenne auto-tuned config: segment_cache=256MB, … table=table_c …   <- identical repeat
00:23:06  Cayenne auto-tuned config: segment_cache=256MB, … table=table_a …   <- identical repeat
00:23:07  Cayenne auto-tuned config: segment_cache=256MB, … table=table_b …   <- identical repeat
…
00:27:35  Cayenne auto-tuned config: segment_cache=256MB, … table=table_c …   <- still going
```

At 1,098 bytes per line this is ~36 KB of duplicate INFO output in the first few minutes, growing unbounded. It buries the `WARN` explaining why the datasets are unhealthy — the one line an operator is actually looking for.

### Expected

The line is genuinely useful — it is the record of what `auto` resolved to on this host, and it is worth keeping at INFO. But it should be emitted once per unique config.

Suggested: key on `(table_name, hash(resolved_config))` and skip the emit when unchanged. That preserves the two cases worth seeing — first resolution, and a *change* in resolution (adaptive tuning moving a knob, or a config edit after a hot reload) — while collapsing the retry storm. A plain "log once per table" would be simpler but would hide a genuine re-tune.

Worth considering separately: whether a deterministic config error should be retried unboundedly at all, since no amount of retrying will resolve it without operator action. Filing that observation here only as context for why the duplication is so visible — happy to split it out.

### Version

`v2.2.0` nightly (2026-07-28). Present on trunk.

### Repro

Any spicepod with a Cayenne-accelerated dataset that fails validation at load will show it. Minimal case — the declared time column's type disagrees with `time_format`, so the dataset never becomes ready and retries forever:

```yaml
version: v1
kind: Spicepod
name: repro

datasets:
  - from: cdc:table_a
    name: table_a
    time_column: ts
    # omitted on purpose: time_format defaults to `timestamp`, which does not
    # match the timezone-carrying column below, so load fails deterministically
    acceleration:
      enabled: true
      engine: cayenne
      mode: file
      refresh_mode: changes
      primary_key: id
      on_conflict:
        id: upsert
    columns:
      - name: id
        type: text
        nullable: false
      - name: ts
        type: Timestamp(Microsecond, UTC)
```

Start the runtime and watch `spiced::acceleration::cayenne` at INFO. The same config line repeats on every retry.

### PR DESCRIPTION
**Pull request by claudespice**
## Summary

`Cayenne auto-tuned config: …` is emitted from `get_vortex_config_with_footer_cache`, which runs on every `create_external_table` / `create_cayenne_table_provider`. Nothing deduped on the resolved configuration, so the line was a function of *how many times a provider was built* rather than *how many distinct configurations exist*.

Dataset initialization retries with unbounded Fibonacci backoff (`crates/runtime/src/init/dataset.rs`) and only `Error::PermanentDatasetFailure` short-circuits, so a dataset that fails a deterministic check is rebuilt for the life of the process — and re-emits its ~1.1 KB config line with every attempt. The reported case produced 33 emissions for 3 tables (3 unique configurations) in the first 4.5 minutes and kept going. The cost is not the bytes: the repeated line buries the `WARN` that explains why the dataset is unhealthy, which is the one line an operator is looking for.

Fixes #12330

## Changes

- `auto_tuned_config_fingerprint` hashes everything the line prints for one table. It hashes the *rendered* values rather than a hand-listed tuple of typed fields, so `{config:?}` covers every `VortexConfig` knob and a knob added to the line later cannot silently fall out of the key and start hiding re-tunes. The calibration measurements on the hardware profile (`data_perf`, `metastore_perf`) are excluded — the line does not print them, and they still reach the fingerprint through the knobs they resolved.
- `auto_tuned_config_is_newly_resolved` records the last fingerprint seen per table and reports only when it is new or changed. Keying on the resolution rather than logging once per table keeps the two cases worth seeing: the first resolution, and a genuine re-tune (adaptive tuning moving a knob, or a configuration edit after a hot reload).
- The `config_warnings` loop moved inside the same gate. Those warnings are a pure function of the same `config` and `hw.cores`, so they repeated in exactly the same storm — and a misconfigured dataset is precisely the one that retries forever, so the reported scenario emits both. One fingerprint covers both emits.

Behaviour on a healthy runtime is unchanged: every dataset still reports its resolution exactly once at startup.

## Scope

Two adjacent emits are deliberately left alone and are not part of this fix: the `cayenne_goal_* set: enabling adaptive tuning` INFO and the `cayenne_tuning: adaptive is in preview` WARN. Both fire only under the adaptive-tuning preview opt-in, and both are emitted from inside the tuning-resolution branch *before* the configuration is final, so they cannot share this fingerprint — gating them needs a different key and is a separate change.

The reporter also raised, as context rather than as part of this report, whether a deterministic configuration error should be retried unboundedly at all. That is the amplifier rather than the duplication, and it is filed separately as #12339.

Existing `warn_once!` / `OnceTracer` was considered and does not fit: it emits a bare `warn!` with no structured fields or custom target, it retains every distinct message string forever, and its never-repeat semantics would suppress the re-tune this line exists to record.

## Test plan

- `make lint`
- `make test`

New unit tests in `crates/runtime/src/dataaccelerator/cayenne/mod.rs`:

- `auto_tuned_config_is_reported_once_per_resolution` — first resolution reports, identical re-resolutions do not, a changed resolution reports again and then settles, reverting to an earlier resolution still reports, and tables are independent.
- `auto_tuned_config_fingerprint_covers_the_logged_values_only` — the fingerprint is deterministic; a knob, the host basis, the inferred workload signals and the table name each change it; a knob the line does not print still participates (over-reporting is the safe direction); a measured storage rate does not.
- `auto_tuned_config_line_survives_a_provider_rebuild_storm` — the regression test. Three provider constructions for one dataset, through the production `get_vortex_config` path with the emitted output captured, must produce exactly one config line and one configuration warning; a subsequent re-tune must report both again.

The regression test was verified load-bearing: reverting the gate to always-emit fails it deterministically with `left: 3, right: 1`, which is the reported failure mode.

## Checklist

- [x] Regression test fails on the old code and passes on the new
- [x] Edge cases covered: re-tune, revert to an earlier resolution, per-table isolation, unprinted knobs, excluded measurements
- [x] No behaviour change on a healthy runtime
- [x] No new dependency
- [x] Deferred item filed separately and linked

### CODE REVIEW

**Comment by claudespice** [comment]
@copilot review

File: crates/runtime/src/dataaccelerator/cayenne/mod.rs [hunk 1]
```diff
@@ -646,6 +646,68 @@ fn warn_if_low_disk_blocking(label: &str, path: &str) {
...
+    hw: &autotune::HardwareProfile,
+    workload: &autotune::WorkloadProfile,
+    config: &cayenne::metadata::VortexConfig,
+) -> u64 {
+    use std::hash::{DefaultHasher, Hash, Hasher};
+
+    let mut hasher = DefaultHasher::new();
+    format!(
+        "{table_name}|{cores}|{total_mem_bytes}|{data_storage:?}|{metastore_storage:?}|\
+         {row_count:?}|{table_bytes:?}|{schema_present}|{has_primary_key}|{is_upsert}|{config:?}",
```
[the remark this answers is not in the document]
Comment (claudespice) [reply crates/runtime/src/dataaccelerator/cayenne/mod.rs:671]: Keeping `{:?}` here — the drift this describes can't happen with these two fields, and where a residual risk exists `Debug` is the safer of the two.

`data_storage`/`metastore_storage` are both `ResolvedAccelerationStorage`, a fieldless enum (`LocalSsd`/`Ebs`/`Tmpfs`/`Unknown`) with a derived `Debug` and a hand-written `Display` in `crates/runtime/src/dataaccelerator/storage.rs`. Both renderings are injective over those four variants, so `{:?}` and `{}` change under exactly the same conditions: the fingerprint moves iff the value moves, which is what the gate needs.

The failure mode left over is a *future* `Display` edit that `Debug` doesn't follow. That one isn't reachable either — `LOGGED` is a process-local `LazyLock<Mutex<HashMap<..>>>`, so a `Display` change arrives in a new binary and therefore a new process, where every table re-emits on its first resolution anyway.

And if the two ever do diverge, `Debug` is the direction you want: a derived `Debug` distinguishes at least as many values as a hand-written `Display`, which is free to collapse variants. Hashing `Debug` can over-report a re-tune; hashing `Display` could miss one. That's the same trade the `{config:?}` field makes deliberately, and the test asserts it (`stream_publish_interval_ms`, a knob the line never prints, still moves the key).

I've taken the doc-comment wording as the fair hit — "the rendered values" reads as a promise of byte-identity with the line, when what it means is "keyed on everything the line depends on". I'll tighten that phrasing if this PR turns for another reason; it isn't worth re-arming a sign-off on its own.

File: crates/runtime/src/dataaccelerator/cayenne/mod.rs [hunk 2]
```diff
@@ -4718,6 +4789,220 @@ mod tests {
...
+            baseline,
+            auto_tuned_config_fingerprint("t", &probed, &workload, &config),
+            "a measured storage rate is not part of the line and must not re-report it"
+        );
+    }
+
+    /// Regression test for #12330: dataset initialization retries with unbounded
+    /// backoff and rebuilds the table provider on every attempt, so a dataset that
+    /// never loads used to repeat its ~1 KB config line forever.
+    #[tokio::test]
```
[the remark this answers is not in the document]
Comment (claudespice) [reply crates/runtime/src/dataaccelerator/cayenne/mod.rs:4929]: The premise is inverted: `#[tokio::test]` defaults to `flavor = "current_thread"`, not multi-thread. It's `#[tokio::main]` that defaults to the multi-thread runtime — the test macro deliberately goes the other way so a test is deterministic unless it opts in with `#[tokio::test(flavor = "multi_thread")]`.

This test takes no `flavor` argument, so the whole body and every future it awaits are polled on the single calling thread — the same thread `set_default` installed the subscriber on. Capture is complete, and the isolation the test comment claims (concurrent tests can't contribute to the count) holds for the same reason.

Forcing `current_thread` explicitly would be a no-op, and switching to a global subscriber would make it strictly worse: `set_global_default` is process-wide and one-shot, so it would capture other tests' emits in the same binary and break the exact-count assertions this test is built on.

Leaving it as is.

**Review by lukekim — APPROVED** [review]

**Review by lukekim — APPROVED** [review]
